src/S3/ApplyChecksumMiddleware.php

```diff
@@ -116,32 +121,114 @@ public function __invoke(
+     * @param bool $checksumRequired
+     * @param string|null $requestedAlgorithm
+     *
+     * @return bool
+     */
+    private function shouldAddChecksum(
+        string $mode,
+        bool $checksumRequired,
+        ?Shape $checksumMember,
+        ?string $requestedAlgorithm
+    ): bool
+    {
+        return ($mode === 'when_supported' && $checksumMember)
+            || ($mode === 'when_required'
+                && ($checksumRequired || ($checksumMember && $requestedAlgorithm)));
+    }
+
+    /**
+     * @param Shape|null $checksumMember
+     * @param string|null $requestedAlgorithm
+     * @param string|null $checksumMemberName
+     *
+     * @return string
+     */
+    private function determineChecksumAlgorithm(
+        array $supportedAlgorithms,
+        ?string $requestedAlgorithm,
+        ?string $checksumMemberName
+    ): string
+    {
+        $algorithm = self::DEFAULT_ALGORITHM;
+
+        if ($requestedAlgorithm) {
+            $requestedAlgorithm = strtolower($requestedAlgorithm);
+            if (!in_array($requestedAlgorithm, $supportedAlgorithms)) {
+                throw new InvalidArgumentException(
+                    "Unsupported algorithm supplied for input variable {$checksumMemberName}. " .
+                    "Supported checksums for this operation include: "
+                    . implode(", ", $supportedAlgorithms) . "."
+                );
+            }
+            $algorithm = $requestedAlgorithm;
+        }
+
+        return $algorithm;
+    }
+
     /**
      * @param string $requestedAlgorithm
      * @param RequestInterface $request
      * @param StreamInterface $body
+     *
      * @return RequestInterface
      */
     private function addAlgorithmHeader(
         string $requestedAlgorithm,
         RequestInterface $request,
         StreamInterface $body
-    ) {
+    ): RequestInterface
+    {
         $headerName = "x-amz-checksum-{$requestedAlgorithm}";
         if (!$request->hasHeader($headerName)) {
-            $encoded = $this->getEncodedValue($requestedAlgorithm, $body);
+            $encoded = self::getEncodedValue($requestedAlgorithm, $body);
             $request = $request->withHeader($headerName, $encoded);
         }
+
         return $request;
     }
 
     /**
-     * @param CommandInterface $command
-     * @return bool
+     * @param string $requestedAlgorithm
+     * @param RequestInterface $request
+     * @param StreamInterface $body
+     *
+     * @return RequestInterface
      */
-    private function isS3Express(CommandInterface $command): bool
+    private function hasAlgorithmHeader(
+        RequestInterface $request,
+        array $supportedAlgorithms
+    ): bool
     {
-        return isset($command['@context']['signing_service'])
-            && $command['@context']['signing_service'] === 's3express';
+        foreach ($supportedAlgorithms as $supportedAlgorithm) {
+            if ($request->hasHeader('x-amz-checksum-' . $supportedAlgorithm)) {
```

Comment:
It should not be validated against our supported algorithms. I know it is odd, but according to the SEP:
> If the user set a checksum header field (i.e. any header prefixed with x-amz-checksum-), then no calculation is necessary. Go to END.

Maybe we should investigate to see if this is the intended behavior here.

Reply:
I figured this would be the safest way to interpret it since something completely incorrect like `x-amz-checksum-algorithm-foo` would also meet that criteria.  I'll get some confirmation on that

Reply:
Looks like we shouldn't validate against anything.  Updated.